xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -494,6 +574,124 @@ private String processRouteConfig(RouteConfiguration config) {
+      if (!adsStream.cdsResourceNames.contains(clusterName)) {
+        continue;
+      }
+      ClusterUpdate.Builder updateBuilder = ClusterUpdate.newBuilder();
+      updateBuilder.setClusterName(clusterName);
+      // The type field must be set to EDS.
+      if (!cluster.getType().equals(DiscoveryType.EDS)) {
+        errorMessage = "Cluster [" + clusterName + "]: only EDS discovery type is supported "
+            + "in gRPC.";
+        break;
+      }
+      // In the eds_cluster_config field, the eds_config field must be set to indicate to
+      // use EDS (must be set to use ADS).
+      EdsClusterConfig edsClusterConfig = cluster.getEdsClusterConfig();
+      if (!edsClusterConfig.hasEdsConfig() || !edsClusterConfig.getEdsConfig().hasAds()) {
+        errorMessage = "Cluster [" + clusterName + "]: field eds_cluster_config must be set to "
+            + "indicate to use EDS over ADS.";
+        break;
+      }
+      // If the service_name field is set, that value will be used for the EDS request
+      // instead of the cluster name (default).
+      if (!edsClusterConfig.getServiceName().isEmpty()) {
+        updateBuilder.setEdsServiceName(edsClusterConfig.getServiceName());
+        edsServices.add(edsClusterConfig.getServiceName());
+      } else {
+        edsServices.add(clusterName);
+      }
+      // The lb_policy field must be set to ROUND_ROBIN.
+      if (!cluster.getLbPolicy().equals(LbPolicy.ROUND_ROBIN)) {
+        errorMessage = "Cluster [" + clusterName + "]: only round robin load balancing policy is "
+            + "supported in gRPC.";
+        break;
+      }
+      updateBuilder.setLbPolicy("round_robin");
+      // If the lrs_server field is set, it must have its self field set, in which case the
+      // client should use LRS for load reporting. Otherwise (the lrs_server field is not set),
+      // LRS load reporting will be disabled.
+      if (cluster.hasLrsServer()) {
+        if (!cluster.getLrsServer().hasSelf()) {
+          errorMessage = "Cluster [" + clusterName + "]: only support enabling LRS for the same "
+              + "management server.";
+          break;
+        }
+        updateBuilder.setEnableLrs(true);
+        updateBuilder.setLrsServerName(serverUri);
+      } else {
+        updateBuilder.setEnableLrs(false);
+      }
+      clusterUpdates.put(clusterName, updateBuilder.build());
+    }
+    if (errorMessage != null) {
+      adsStream.sendNackRequest(ADS_TYPE_URL_CDS, adsStream.cdsResourceNames, errorMessage);
+      return;
+    }
+    adsStream.sendAckRequest(ADS_TYPE_URL_CDS, adsStream.cdsResourceNames,
+        cdsResponse.getVersionInfo());
+
+    // Update local CDS cache with data in this response.
+    clusterNamesToClusterUpdates.clear();
+    clusterNamesToClusterUpdates.putAll(clusterUpdates);
+
+    // Remove EDS cache entries for ClusterLoadAssignments not referenced by this CDS response.
+    clusterNamesToEndpointUpdates.keySet().retainAll(edsServices);
```

Comment:
This will be taken care of automatically because the CdsLoadBalancer will cancel the unused cluster watcher and shutdown the corresponding EdsLoadBalancer, which in turn will cancel the endpoint watcher.

Reply:
What? This is for cached data. Nothing to do with watchers.

Reply:
That's `cancelEndpointDataWatch()`'s job, which is taken care of. The Cds response handle should not touch eds cached updates.

Reply:
No, EDS is incremental, if some `ClusterLoadAssignment` is removed, this is reflected in CDS response. This is how the protocol works. Check design doc.

Reply:
I know EDS is incremental, I'm not talking about handling EDS response at all. When CDS resource is removed, it will shutdown the corresponding EdsLoadBalancer, and all EndpointWatchers will be cancelled. In your `cancelEndpointDataWatch()`, you have 

```java
  if (watchers.isEmpty()) {
      logger.log(Level.FINE, "Stop watching endpoints in cluster {0}", clusterName);
      endpointWatchers.remove(clusterName);
      // Remove the corresponding EDS cache entry.
      clusterNamesToEndpointUpdates.remove(clusterName);
```
That's removing cached update.

Reply:
This is the first place of handling CDS response, caches are maintained internally in XdsClient. Implementation should not rely on assumptions of watcher implementations.

Reply:
EndpointUpdate caches are maintained internally in `cancelEndpointDataWatch` in XdsClient. Only EDS part of XdsClient should maintain `clusterNamesToEndpointUpdates`. CDS part of XdsClient should not touch `clusterNamesToEndpointUpdates`.

Reply:
> EndpointUpdate caches are maintained internally in `cancelEndpointDataWatch` in XdsClient.
You should NOT implement something based on assumptions of how clients/users of this class will use it.

As I said, `ClusterLoadAssignment` removal is reflected in CDS responses. Therefore, if a CDS response no longer references some cluster, that indicates the corresponding `ClusterLoadAssignment` no longer exists. You should not assume giving an updated `ClusterUpdate` to CDS policy will always have a consequent `cancelEndpointDataWatch` be called.

> CDS part of XdsClient should not touch `clusterNamesToEndpointUpdates`.
This is how the protocol works. CDS indicates something about EDS.

Reply:
We didn't reach an agreement. @ejona86 and me didn't convince @voidzcy and vice versa. But as this is not so critical, I don't want this PR to be blocked because this is unresolved. I can accept it as-is. Let proceed.

Reply:
Did I explain this to you in the later conversation and seems you somewhat agreed with what I thought? But, whatever, let's proceed to make things work with minimum requirement first.